File: frontend/src/types/recurring.ts
import type { TransactionType } from "./transaction";

export type RecurringFrequency = "daily" | "weekly" | "monthly" | "yearly";

// Matches the backend's RecurringOut schema exactly - monthly_equivalent
// is computed server-side at read time, never stored.
export interface RecurringTransaction {
  id: number;
  category_id: number | null;
  category_name: string | null;
  category_icon: string | null;
  amount: number;
  type: TransactionType;
  description: string;
  frequency: RecurringFrequency;
  next_date: string; // ISO date string
  active: boolean;
  monthly_equivalent: number;
  created_at: string;
  updated_at: string;
}

export interface RecurringInput {
  category_id: number | null;
  amount: number;
  type: TransactionType;
  description: string;
  frequency: RecurringFrequency;
  next_date: string;
  active: boolean;
}

File: backend/app/schemas/recurring.py
from datetime import date as date_type
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

RecurringType = Literal["income", "expense"]
RecurringFrequency = Literal["daily", "weekly", "monthly", "yearly"]


class RecurringBase(BaseModel):
    category_id: int | None = None
    amount: float = Field(..., gt=0, description="Amount must be greater than zero")
    type: RecurringType
    description: str = Field(..., min_length=1, max_length=255)
    frequency: RecurringFrequency
    next_date: date_type
    active: bool = True


class RecurringCreate(RecurringBase):
    pass


class RecurringUpdate(RecurringBase):
    pass


class RecurringOut(RecurringBase):
    """
    monthly_equivalent is computed at read time (never stored) - it's what
    this commitment works out to per month regardless of its actual
    frequency, so a weekly and a yearly commitment can be compared and
    summed on the same basis (PRD section 26).
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    category_name: str | None
    category_icon: str | None
    monthly_equivalent: float
    created_at: datetime
    updated_at: datetime
